src/renderers/shared/fiber/ReactFiberScheduler.js

```diff
@@ -498,6 +479,75 @@ module.exports = function<T, P, I, TI, C>(config : HostConfig<T, P, I, TI, C>) {
     return root;
   }
 
+  function performSynchronousWorkUnsafe() {
+    // Perform work now
+    nextUnitOfWork = findNextUnitOfWork();
+    while (nextUnitOfWork &&
+           nextPriorityLevel === SynchronousPriority) {
+      nextUnitOfWork = performUnitOfWork(nextUnitOfWork, false);
+
+      if (!nextUnitOfWork && shouldBatchUpdates) {
+        nextUnitOfWork = findNextUnitOfWork();
+      }
+    }
+    if (nextUnitOfWork) {
+      if (nextPriorityLevel > AnimationPriority) {
+        scheduleDeferredCallback(performDeferredWork);
+        return;
+      }
+      scheduleAnimationCallback(performAnimationWork);
+    }
+  }
+
+  function performSynchronousWork() {
+    if (useSyncScheduling) {
+      // Start batching updates
+      shouldBatchUpdates = true;
+    }
+    performAndHandleErrors(performSynchronousWorkUnsafe);
+    shouldBatchUpdates = false;
+  }
+
+  function scheduleSynchronousWork(root : FiberRoot) {
+    root.current.pendingWorkPriority = SynchronousPriority;
+
+    if (root.isScheduled) {
+      // If we're already scheduled, we can bail out.
+      return;
+    }
+    root.isScheduled = true;
+    if (lastScheduledRoot) {
+      // Schedule ourselves to the end.
+      lastScheduledRoot.nextScheduledRoot = root;
+      lastScheduledRoot = root;
+    } else {
+      // We're the only work scheduled.
+      nextScheduledRoot = root;
+      lastScheduledRoot = root;
+
+      if (!shouldBatchUpdates) {
+        performSynchronousWork();
+      }
+    }
+  }
+
+  function performAndHandleErrors<A>(fn: (a: A) => void, a: A) {
+    try {
+      fn(a);
```

Comment:
Ideally we should avoid first-class functions because I'd like everything to be able to take advantage of inline expansion - and be easily portable to C. It also helps newcomers to follow code that you don't have to jump back and forth. Not sure if we'll even use first-class functions for scheduling in React Native because we're not tied to rAF/rIC. Might just move that part to the DOM renderer.

It'd probably be better to go in the opposite direction and unify deferred/animation work and simply branch in performWork before the loop.